core/src/main/java/smile/manifold/UMAP.java

```diff
@@ -0,0 +1,846 @@
+                        }
+                    }
+                }
+
+                alpha = initialAlpha * (1.0 - (float) n / (float) nEpochs);
+                logger.debug("completed " + (n + 1) + " epochs optimization on embedding.");
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            logger.debug(e.getMessage());
+        }
+    }
+
+    /**
+     * return value from given hashmaps, possible 0 if no mapping existing for
+     * given i->j->val
+     * 
+     * @param values
+     *            hashmaps to find value
+     * @param i
+     *            first mapping index
+     * @param j
+     *            second mapping index
+     * @return value for given mapping indexes: i->j->val
+     */
+    private static double getDoubleValueFrom(Map<Integer, Map<Integer, Double>> values, int i, int j) {
+        if (values.get(i) != null) {
+            Map<Integer, Double> iVals = values.get(i);
+            if (iVals.get(j) != null) {
+                return iVals.get(j);
+            } else {
+                return 0;
+            }
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+     * Given a set of weights and number of epochs generate the number of epochs
+     * per sample for each weight.
+     * 
+     * @param weight
+     *            The weights of how much we wish to sample each 1-simplex as
+     *            {@link SparseMatrix}.
+     * @param nEpochs
+     *            The total number of epochs we want to train for.
+     * @return An array of number of epochs per sample, one for each 1-simplex
+     *         between (ith, jth) sample point.
+     */
+    private static Map<Integer, Map<Integer, Double>> makeEpochsPerSample(SparseMatrix weight, int nEpochs) {
+
+        Map<Integer, Map<Integer, Double>> ret = new HashMap<Integer, Map<Integer, Double>>();
+
+        double maxW = Double.MIN_VALUE;
+        Iterator<SparseMatrix.Entry> nonZeros = weight.iterator();
+        while (nonZeros != null && nonZeros.hasNext()) {
+            SparseMatrix.Entry nonZero = nonZeros.next();
+            if (nonZero != null) {
+                if (nonZero.x > maxW) {
+                    maxW = nonZero.x;
+                }
+            } else {
+                break;
+            }
+        }
+
+        nonZeros = weight.iterator();
+        while (nonZeros != null && nonZeros.hasNext()) {
+            SparseMatrix.Entry nonZero = nonZeros.next();
+            if (nonZero != null) {
+                double nSamples = (double) nEpochs * (nonZero.x / maxW);
+                Map<Integer, Double> weights = ret.get(nonZero.i);
+                if (weights == null) {
+                    weights = new HashMap<Integer, Double>();
+                    ret.put(nonZero.i, weights);
+                }
+                if (nSamples > 0) {
+                    weights.put(nonZero.j, (double) nEpochs / nSamples);
```

Comment:
This doesn't make sense to me with the formula in Line 810. Why not just `maxW / x`?